tests/test_connections.py

```diff
@@ -242,3 +244,119 @@ def test_post_request_fail():
+    assert user.commands[3]["remove"][GroupTypes.usergroup.name] == remove_groups[10:]
+
+
+def test_large_group_action_split():
+    """
+    Ensure that very large group lists (100+) will be handled appropriately
+    Connection.execute_multiple splits commands and splits actions
+    Result should be 2 actions, even though we only created one action
+    :return:
+    """
+    with mock.patch("umapi_client.connection.requests.Session.post") as mock_post:
+        mock_post.return_value = MockResponse(200, {"result": "success"})
+        conn = Connection(**mock_connection_params)
+
+        group_prefix = "G"
+        add_groups = [group_prefix+six.text_type(n+1) for n in range(0, 150)]
+        user = UserAction(id_type=IdentityTypes.enterpriseID, email="user@example.com")
+        user.add_to_groups(groups=add_groups, group_type=GroupTypes.usergroup)
+        assert conn.execute_single(user, immediate=True) == (0, 2, 2)
+
+
+def test_group_size_limit():
+    """
+    Test with different 'throttle_groups' value, which governs the max size of the group list before commands are split
+    :return:
+    """
+    with mock.patch("umapi_client.connection.requests.Session.post") as mock_post:
+        mock_post.return_value = MockResponse(200, {"result": "success"})
+        params = mock_connection_params
+        params['throttle_groups'] = 5
+        conn = Connection(**params)
+
+        group_prefix = "G"
+        add_groups = [group_prefix+six.text_type(n+1) for n in range(0, 150)]
+        user = UserAction(id_type=IdentityTypes.enterpriseID, email="user@example.com")
+        user.add_to_groups(groups=add_groups, group_type=GroupTypes.usergroup)
+        assert conn.execute_single(user, immediate=True) == (0, 3, 3)
+
+
+def test_split_add_user():
+    """
+    Make sure split doesn't do anything when we have a non-add/remove group action
+    :return:
+    """
+    user = UserAction(id_type=IdentityTypes.enterpriseID, email="user@example.com")
+    user.create(first_name="Example", last_name="User", country="US", email="user@example.com")
+    assert user.maybe_split_groups(10) is False
+
+
+def test_split_role_assignment():
+    group_prefix = "G"
+    add_groups = [group_prefix+six.text_type(n+1) for n in range(0, 25)]
+    user = UserAction(id_type=IdentityTypes.enterpriseID, email="user@example.com")
+    user.add_role(groups=add_groups, role_type=RoleTypes.admin)
+    assert user.maybe_split_groups(10) is True
+    assert len(user.commands) == 3
+
+
+def test_no_group_split():
+    """
+    maybe_split should return false if nothing was split
+    :return:
+    """
+    group_prefix = "G"
+    add_groups = [group_prefix+six.text_type(n+1) for n in range(0, 5)]
+    user = UserAction(id_type=IdentityTypes.enterpriseID, email="user@example.com")
+    user.add_to_groups(groups=add_groups, group_type=GroupTypes.usergroup)
+    assert user.maybe_split_groups(10) is False
+    assert len(user.commands) == 1
```

Comment:
What happened to the test we discussed that had multiple different group types being added in one command, where one had 30, one had 15, and one had 5 groups?  You were going to make sure we only added the ones that needed to be added to the second and third commands?

Reply:
Oh, good point, I forgot there was that limitation in the UserAction interface.  (I put the limitation in because there really isn't any difference in the meaning of the group types, so there was no real need for someone to use multiple.)  Let's do the test anyway, constructing the list by hand for now, since the code looks for and handles that case we should at least test it.  Probably a comment explaining why the list is constructed by hand would be useful.

Reply:
I'll add a similar test case that will test multiple group types in a single step.  Since my latest revision doesn't handle multiple step keys, it won't test for that anymore.  (in fact, I need to error check multiple keys)

Reply:
Yes, error checking multiple keys would be nice, but not required.  A comment to that effect is probably enough so as to avoid spurious breakage of clients who are getting away with it currently...